File: Top-Interview-Questions/Array/Easy-Collection/217-contains-duplicate.py
from collections import defaultdict
class Solution:
    # Sol 1
    def containsDuplicate(self, nums: List[int]) -> bool:
        m = defaultdict(int)
        for num in nums:
            if m[num]:
                return True
            m[num]+=1
        return False
    # Time Complexity: O(n)
    # Space Complexity: O(n), used a map, same memory as the array

File: Top-Interview-Questions/Array/Easy-Collection/217-containsDuplicate.cpp
class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
        /*
         * Time Complexity: O(n) 
         * Space Complexity: O(n)  
         */

        int n = nums.size();
        
        if(n == 0 || n == 1)
            return false;
        
        map<int, bool> a;

        for(int i = 0; i < n; i++){
            
            if(a[nums[i]] == true){
                return true;
            }
            
            a[nums[i]] = true;
        }
        
        return false;
    }
};